### ISSUE TITLE
SIGABRT from czmq

### ISSUE THREAD

**Issue by mg964** [zeromq/czmq#2166]
Trying to add a simple ZMQ-based request/reply mechanism to an existing daemon (twampd) that makes heavy use of forks (children) and signals. Essentially a child is used to process an inbound network request and a signal is used to terminate/cleanup the child.
However having added some really basic ZMQ calls, the daemon now terminates with a SIGABRT. The ZMQ addition is simply this:

	const char * path ="ipc:///<path>";
	zsock_t *s = zsock_new_req(path);
	zsock_destroy(&s);

If I now trigger a new connection to the daemon, the child gets created and all looks OK. The daemon then terminates with the ABRT. GBD says this:

(gdb) cont
Continuing.
[Detaching after fork from child process 24288]

Thread 2 "ZMQbg/0" received signal SIGABRT, Aborted.
[Switching to Thread 0x7fab65f49700 (LWP 23983)]
0x00007fab66e637bb in raise () from /lib/x86_64-linux-gnu/libc.so.6
(gdb) bt
#0  0x00007fab66e637bb in raise () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007fab66e4e535 in abort () from /lib/x86_64-linux-gnu/libc.so.6
#2  0x00007fab67190773 in ?? () from /lib/x86_64-linux-gnu/libzmq.so.5
#3  0x00007fab671af436 in ?? () from /lib/x86_64-linux-gnu/libzmq.so.5
#4  0x00007fab671ccab9 in ?? () from /lib/x86_64-linux-gnu/libzmq.so.5
#5  0x00007fab671a696c in ?? () from /lib/x86_64-linux-gnu/libzmq.so.5
#6  0x00007fab671e8cf4 in ?? () from /lib/x86_64-linux-gnu/libzmq.so.5
#7  0x00007fab6676cfa3 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#8  0x00007fab66f254cf in clone () from /lib/x86_64-linux-gnu/libc.so.6
(gdb) i thread
  Id   Target Id                                   Frame 
  1    Thread 0x7fab65f64c80 (LWP 23982) "twampd"  0x00007fab66f1a819 in poll () from /lib/x86_64-linux-gnu/libc.so.6
* 2    Thread 0x7fab65f49700 (LWP 23983) "ZMQbg/0" 0x00007fab66e637bb in raise () from /lib/x86_64-linux-gnu/libc.so.6
  3    Thread 0x7fab65748700 (LWP 23984) "ZMQbg/1" 0x00007fab66f257ef in epoll_wait () from /lib/x86_64-linux-gnu/libc.so.6
(gdb) thread 1
[Switching to thread 1 (Thread 0x7fab65f64c80 (LWP 23982))]
#0  0x00007fab66f1a819 in poll () from /lib/x86_64-linux-gnu/libc.so.6
(gdb) bt
#0  0x00007fab66f1a819 in poll () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x0000561dc856ca80 in main (argc=0, argv=0x7ffca66b0890) at owampd.c:2206
(gdb)

How can I figure out why the ZMQ thread is issuing the SIGABRT? 

(czmq version: libczmq.so.4.2.0)

**Comment by bluca** [comment]
Please install the -dbgsym packages to get debugging symbols and a check the backtrace again

**Comment by mg964** [comment]
Hey Luca, where can I find the latest symbols? I have libzmq5 "Version: 4.3.1-4+deb10u2" installed, yet the only dbg package I can find is "libzmq5-dbg_4.2.1-4+deb9u3_amd64.deb" (https://packages.debian.org/stretch/amd64/libzmq5-dbg/download). 

I can see a libzmq5-dbgsym (4.3.4-1) on debports, but can't find the amd64 package.
Mark

**Comment by bluca** [comment]
https://wiki.debian.org/HowToGetABacktrace

**Comment by mg964** [comment]
Right, finally managed to get the correct symbols loaded:
```
(gdb) i thread
  Id   Target Id                                   Frame 
  1    Thread 0x7fe2c3570c80 (LWP 28667) "twampd"  0x00007fe2c4526819 in __GI___poll (fds=0x560f58247f00, nfds=2, 
    timeout=-1) at ../sysdeps/unix/sysv/linux/poll.c:29
* 2    Thread 0x7fe2c3555700 (LWP 28668) "ZMQbg/0" __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
  3    Thread 0x7fe2c2d54700 (LWP 28669) "ZMQbg/1" 0x00007fe2c45317ef in epoll_wait (epfd=11, 
    events=events@entry=0x7fe2c2d52cb0, maxevents=maxevents@entry=256, timeout=timeout@entry=-1)
    at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007fe2c445a535 in __GI_abort () at abort.c:79
#2  0x00007fe2c479c773 in zmq::zmq_abort (errmsg_=errmsg_@entry=0x7fe2c4806a83 "ok") at src/err.cpp:88
#3  0x00007fe2c47bb436 in zmq::mailbox_t::recv (this=this@entry=0x560f5823d570, cmd_=cmd_@entry=0x7fe2c3553bc0, 
    timeout_=timeout_@entry=0) at src/mailbox.cpp:99
#4  0x00007fe2c47d8ab9 in zmq::reaper_t::in_event (this=0x560f5823d550) at src/reaper.cpp:96
#5  0x00007fe2c47b296c in zmq::epoll_t::loop (this=0x560f5823d800) at src/epoll.cpp:206
#6  0x00007fe2c47f4cf4 in thread_routine (arg_=0x560f5823d858) at src/thread.cpp:182
#7  0x00007fe2c3d78fa3 in start_thread (arg=<optimized out>) at pthread_create.c:486
#8  0x00007fe2c45314cf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
(gdb)
```
Looks like I'm missing a read of some form, but the code is trivial. I'll keep poking about in the main twampd code, but any insight into what ZMQ is doing would be helpful.

**Comment by bluca** [comment]
Is this really what you have:

```
const char * path ="ipc:///<path>";
zsock_t *s = zsock_new_req(path);
zsock_destroy(&s);
```

Are you checking that s was created successfully?

**Comment by mg964** [comment]
The additional code is now this:
	const char *path = "ipc:///var/run/test.control";
	zsock_t *s = zsock_new_req(path);

	assert(s != NULL);
	OWPError(ctx, OWPErrDEBUG, OWPErrINVALID,
		 "INIT-ZMQ path %s zsock_is %d zsock_type %s",
		 path, zsock_is(s), zsock_type_str(s));

	zsock_destroy(&s);

The logging produces this:
```
Mar 02 09:29:05 vm-twamp0-3 twampd[32702]: INIT-ZMQ path ipc:///var/run/test.control zsock_is 1 zsock_type REQ
Mar 02 09:29:26 vm-twamp0-3 twampd[32705]: Connection to ([10.3.1.3]:862) from ([10.1.2.1]:36218)
Mar 02 09:29:26 vm-twamp0-3 twampd[32705]: ControlSession([10.3.1.3]:862) accepted from userid(nil):([10.1.2.1]:36218)
Mar 02 09:29:26 vm-twamp0-3 twampd[32705]: ResReq ALLOWED: open:request:test_sessions = 1 (result = 1, limit = 8)
Mar 02 09:29:26 vm-twamp0-3 twampd[32705]: _OWPEndpointInitHook()
Mar 02 09:29:26 vm-twamp0-3 twampd[32705]: _OWPEndpointStart()
Mar 02 09:29:26 vm-twamp0-3 twampd[32705]: StartSessions 1 sessions
Mar 02 09:29:38 vm-twamp0-3 twampd[32705]: _OWPEndpointStop()
Mar 02 09:29:38 vm-twamp0-3 systemd[1]: twamp-server.service: Main process exited, code=killed, status=6/ABRT
Mar 02 09:29:44 vm-twamp0-3 systemd[1]: twamp-server.service: State 'stop-sigterm' timed out. Killing.
Mar 02 09:29:44 vm-twamp0-3 systemd[1]: twamp-server.service: Killing process 32705 (twampd) with signal SIGKILL.
Mar 02 09:29:44 vm-twamp0-3 systemd[1]: twamp-server.service: Killing process 32706 (twampd) with signal SIGKILL.
Mar 02 09:29:44 vm-twamp0-3 systemd[1]: twamp-server.service: Failed with result 'signal'.
```
Note how systemd had to issue an explicit kill in order to cleanup following the ABRT. 

Its got to have something to do with signals. AFAICT SIGUSR1 is used to start a test session and SIGUSR2 is used to terminate the associated child process. Havn't figured out what the others are for; these are the signals used by twampd:

    if(     (sigaction(SIGPIPE,&ignact,NULL) != 0)  ||
            (sigaction(SIGTERM,&setact,NULL) != 0)  ||
            (sigaction(SIGUSR1,&setact,NULL) != 0)  ||
            (sigaction(SIGUSR2,&setact,NULL) != 0)  ||
            (sigaction(SIGINT,&setact,NULL) != 0)   ||
            (sigaction(SIGHUP,&setact,NULL) != 0)   ||
            (sigaction(SIGCHLD,&setact,NULL) != 0)  ||
            (sigaction(SIGALRM,&setact,NULL) != 0)  ){

Could any of these somehow conflict with ZMQ?

**Comment by bluca** [comment]
only term/int - see zsys_handler_set/zsys_handler_reset

**Comment by mg964** [comment]
Not sure what you had in mind, but adding "zsys_handler_set(NULL)" made no difference.

The crash is from here (zmq::mailbox_t::recv):
```
    //  Get a command.
    const bool ok = _cpipe.read (cmd_);
    zmq_assert (ok);
    return 0;
```
Why does it think there is something to be read? Who/what is supposed to be on the other end of this pipe?

Don't know where to go from here - feels like this twampd code is somehow incompatible with ZMQ.

**Comment by bluca** [comment]
> Not sure what you had in mind, but adding "zsys_handler_set(NULL)" made no difference.

You mentioned signal handling, and that's the extent of it - catching int/term

> The crash is from here (zmq::mailbox_t::recv):
> 
> ```
>     //  Get a command.
>     const bool ok = _cpipe.read (cmd_);
>     zmq_assert (ok);
>     return 0;
> ```
> 
> Why does it think there is something to be read? Who/what is supposed to be on the other end of this pipe?
> 
> Don't know where to go from here - feels like this twampd code is somehow incompatible with ZMQ.

Looking at the timestamps from your logs, it's probably happening in the atexit handler - I wouldn't focus too much on the mailbox itself, that's just a symptom, the issue is that there's some corruption somewhere and somehow an invalid socket is still being processed. Is this the only usage of czmq? Are multiple threads involved in any way?

**Comment by mg964** [comment]
> Is this the only usage of czmq? Are multiple threads involved in any way?

Yes. After problems with some initial code, I removed everything to leave the above test sequence.

The process forks to create the daemon and then forks again in order to process the inbound "ping" request (_OWPEndpointInitHook) which is then killed off by the control thread (_OWPEndpointStop). Its at this point that something "upsets" ZMQ. Note that the main control process should remain, waiting for the next ping request.
The above ZMQ test code occurs after the initial (control thread) fork, but I see the same issue if I run the process "interactively" (no daemon fork).

I'll keep playing around, see if I can follow the various threads and signals.

**Comment by bluca** [comment]
> > Is this the only usage of czmq? Are multiple threads involved in any way?
> 
> Yes. After problems with some initial code, I removed everything to leave the above test sequence.
> 
> The process forks to create the daemon and then forks again in order to process the inbound "ping" request (_OWPEndpointInitHook) which is then killed off by the control thread (_OWPEndpointStop). Its at this point that something "upsets" ZMQ. Note that the main control process should remain, waiting for the next ping request.
> The above ZMQ test code occurs after the initial (control thread) fork, but I see the same issue if I run the process "interactively" (no daemon fork).
> 
> I'll keep playing around, see if I can follow the various threads and signals.

How is the forked process killed off?

**Comment by mg964** [comment]
> How is the forked process killed off?

Looks like its sent a SIGUSR2 (12) (or SIGINT in the case of a error termination). SIGUSR1 (10) is used to kick-start the child into accepting frames.

```
Mar 03 09:49:18 vm-twamp0-3 twampd[5992]: StartSessions 1 sessions
Mar 03 09:49:18 vm-twamp0-3 twampd[5993]: _OWPEndpointInitHook(4153) lastsigno 0 sig_nothing_signo 0 sig_catch_signo 10
Mar 03 09:49:31 vm-twamp0-3 twampd[5992]: _OWPEndpointStop()
Mar 03 09:49:31 vm-twamp0-3 twampd[5992]: _OWPEndpointStop() issuing kill(5993, 12)
Mar 03 09:49:31 vm-twamp0-3 systemd[1]: twamp-server.service: Main process exited, code=killed, status=6/ABRT
Mar 03 09:49:37 vm-twamp0-3 systemd[1]: twamp-server.service: State 'stop-sigterm' timed out. Killing.
```

**Comment by bluca** [comment]
Are you sure no czmq api is a called at all anywhere before the final fork?

**Comment by mg964** [comment]
To be 100% certain I ripped out all previous code and am left with just that basic socket create/delete snippet that takes place as "main" gets going. In the failed case there are 3 forks (processes) involved - "main" (daemon), "control (5992)" and "test (5993)".

**Comment by mg964** [comment]
OK now I'm very confused. Replaced the previous snippet with:
```
	zsys_handler_set(NULL);
	void *zctx = zmq_ctx_new ();
	assert (zctx);
	void *socket = zmq_socket (zctx, ZMQ_STREAM);
	assert (socket);
	zmq_close (socket);
	zmq_ctx_destroy (zctx);
	return;
```
The test ping passed fine, no ABRT. Is this use of zmq so radically different to the czmq "wrappers"? Any thoughts or suggestions as where to go now?

**Comment by bluca** [comment]
Yes, the issue is most likely with the atexit handler that removes the context automatically as mentioned earlier - that's why I'm asking about processes, things sometimes get weird if you fork after using these

**Comment by bluca** [comment]
you can find the implementation in zsys.c - the atexit handler goes throught he list of sockets and closes them if still open, and then destroys the context

**Comment by mg964** [comment]
Out of interested I added a call to zsys_shutdown() at the end of the test sequence, no errors. I then went back to the original development code that connected to the target daemon, issued an initial "hello" and waited for the response. Once the zsys_shutdown() call was added to that chunk of code, all worked fine.

So as you indicated, the ABRT has all tied up with the various process forks issued by twamp. Don't know if I can develop the rest of the code using zsys_shutdown(), maybe the only way to proceed.
